File: aip/packages/typescript/src/window/manager.ts
import type { IntegrityCheckpoint } from "../schemas/checkpoint.js";
import type { WindowConfig } from "../schemas/config.js";
import type { WindowSummary } from "../schemas/signal.js";
import type { WindowState } from "./state.js";
import { createWindowState } from "./state.js";
import { MIN_WINDOW_SIZE } from "../constants.js";

export class WindowManager {
  private state: WindowState;
  private readonly config: WindowConfig;

  constructor(config: WindowConfig, sessionId: string) {
    if (config.max_size < MIN_WINDOW_SIZE) {
      throw new Error(
        `Window max_size must be >= ${MIN_WINDOW_SIZE}, got ${config.max_size}`
      );
    }
    this.config = config;
    this.state = createWindowState(sessionId);
  }

  /** Add a checkpoint to the window. Handles eviction per mode. */
  push(checkpoint: IntegrityCheckpoint): void {
    // 1. Check session boundary
    if (checkpoint.session_id !== this.state.session_id) {
      if (this.config.session_boundary === "reset") {
        this.reset();
      }
      this.state.session_id = checkpoint.session_id;
    }

    // 2. Evict stale checkpoints (older than max_age_seconds)
    this.evictStale();

    // 3. If at max_size, handle eviction per mode
    if (this.state.checkpoints.length >= this.config.max_size) {
      if (this.config.mode === "sliding") {
        this.state.checkpoints.shift();
      } else {
        // fixed mode: reset entirely, then insert
        this.state.checkpoints = [];
      }
    }

    // 4. Push checkpoint, increment total_checks
    this.state.checkpoints.push(checkpoint);
    this.state.stats.total_checks++;

    // 5. Recompute window stats
    this.recomputeStats();
  }

  /** Format window for conscience prompt (numbered verdict list) */
  getContext(): string {
    if (this.state.checkpoints.length === 0) {
      return "SESSION CONTEXT: First check in session (no prior context)";
    }
    const lines = this.state.checkpoints.map(
      (cp, i) => `${i + 1}. [${cp.verdict}] ${cp.reasoning_summary}`
    );
    return `SESSION CONTEXT (window: ${this.state.size}/${this.config.max_size}):\n${lines.join("\n")}`;
  }

  /** Clear all checkpoints and reset statistics */
  reset(): void {
    this.state.checkpoints = [];
    this.state.size = 0;
    this.state.stats.total_checks = 0;
    this.state.stats.clear_count = 0;
    this.state.stats.review_count = 0;
    this.state.stats.violation_count = 0;
    this.state.stats.avg_analysis_ms = 0;
  }

  /** Get WindowSummary for IntegritySignal */
  getSummary(): WindowSummary {
    const verdicts = { clear: 0, review_needed: 0, boundary_violation: 0 };
    for (const cp of this.state.checkpoints) {
      verdicts[cp.verdict]++;
    }
    return {
      size: this.state.size,
      max_size: this.config.max_size,
      verdicts,
      integrity_ratio:
        this.state.size > 0 ? verdicts.clear / this.state.size : 1.0,
      drift_alert_active: false, // Set externally by drift detection
    };
  }

  /** Get full window state (for SDK getWindowState()) */
  getState(): WindowState {
    return { ...this.state, checkpoints: [...this.state.checkpoints] };
  }

  /** Get current session ID */
  getSessionId(): string {
    return this.state.session_id;
  }

  /** Evict checkpoints older than max_age_seconds */
  private evictStale(): void {
    const now = Date.now();
    const maxAgeMs = this.config.max_age_seconds * 1000;
    this.state.checkpoints = this.state.checkpoints.filter(
      (cp) => now - Date.parse(cp.timestamp) <= maxAgeMs
    );
  }

  /** Recompute window-level stats from current checkpoints */
  private recomputeStats(): void {
    const cps = this.state.checkpoints;
    this.state.size = cps.length;
    this.state.stats.clear_count = 0;
    this.state.stats.review_count = 0;
    this.state.stats.violation_count = 0;

    let totalMs = 0;
    for (const cp of cps) {
      switch (cp.verdict) {
        case "clear":
          this.state.stats.clear_count++;
          break;
        case "review_needed":
          this.state.stats.review_count++;
          break;
        case "boundary_violation":
          this.state.stats.violation_count++;
          break;
      }
      totalMs += cp.analysis_metadata.analysis_duration_ms;
    }
    this.state.stats.avg_analysis_ms =
      cps.length > 0 ? totalMs / cps.length : 0;
  }
}

File: aip/packages/python/src/aip/window/manager.py
"""Window manager — mirrors packages/typescript/src/window/manager.ts."""

from __future__ import annotations

import copy
import time
from datetime import datetime

from aip.constants import MIN_WINDOW_SIZE
from aip.schemas.checkpoint import IntegrityCheckpoint
from aip.schemas.config import WindowConfig
from aip.schemas.signal import VerdictCounts, WindowSummary
from aip.window.state import WindowState, create_window_state


class WindowManager:
    """Manages the sliding/fixed integrity checkpoint window.

    Handles checkpoint storage, eviction (age + size), session boundaries,
    and produces window summaries for signal construction.
    """

    def __init__(self, config: WindowConfig, session_id: str) -> None:
        if config.max_size < MIN_WINDOW_SIZE:
            raise ValueError(
                f"Window max_size must be >= {MIN_WINDOW_SIZE}, "
                f"got {config.max_size}"
            )
        self._config = config
        self._state = create_window_state(session_id)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def push(self, checkpoint: IntegrityCheckpoint) -> None:
        """Add a checkpoint to the window. Handles eviction per mode."""

        # 1. Check session boundary
        if checkpoint.session_id != self._state.session_id:
            if self._config.session_boundary == "reset":
                self.reset()
            self._state.session_id = checkpoint.session_id

        # 2. Evict stale checkpoints (older than max_age_seconds)
        self._evict_stale()

        # 3. If at max_size, handle eviction per mode
        if len(self._state.checkpoints) >= self._config.max_size:
            if self._config.mode == "sliding":
                self._state.checkpoints.pop(0)
            else:
                # fixed mode: reset entirely, then insert
                self._state.checkpoints = []

        # 4. Push checkpoint, increment total_checks
        self._state.checkpoints.append(checkpoint)
        self._state.stats.total_checks += 1

        # 5. Recompute window stats
        self._recompute_stats()

    def get_context(self) -> str:
        """Format window for conscience prompt (numbered verdict list)."""
        if len(self._state.checkpoints) == 0:
            return "SESSION CONTEXT: First check in session (no prior context)"

        lines = [
            f"{i + 1}. [{cp.verdict}] {cp.reasoning_summary}"
            for i, cp in enumerate(self._state.checkpoints)
        ]
        return (
            f"SESSION CONTEXT (window: {self._state.size}/{self._config.max_size}):\n"
            + "\n".join(lines)
        )

    def reset(self) -> None:
        """Clear all checkpoints and reset statistics."""
        self._state.checkpoints = []
        self._state.size = 0
        self._state.stats.total_checks = 0
        self._state.stats.clear_count = 0
        self._state.stats.review_count = 0
        self._state.stats.violation_count = 0
        self._state.stats.avg_analysis_ms = 0

    def get_summary(self) -> WindowSummary:
        """Get WindowSummary for IntegritySignal."""
        verdicts = VerdictCounts(clear=0, review_needed=0, boundary_violation=0)
        for cp in self._state.checkpoints:
            if cp.verdict == "clear":
                verdicts.clear += 1
            elif cp.verdict == "review_needed":
                verdicts.review_needed += 1
            elif cp.verdict == "boundary_violation":
                verdicts.boundary_violation += 1

        integrity_ratio = (
            verdicts.clear / self._state.size if self._state.size > 0 else 1.0
        )

        return WindowSummary(
            size=self._state.size,
            max_size=self._config.max_size,
            verdicts=verdicts,
            integrity_ratio=integrity_ratio,
            drift_alert_active=False,  # Set externally by drift detection
        )

    def get_state(self) -> WindowState:
        """Get full window state (for SDK get_window_state())."""
        return WindowState(
            checkpoints=list(self._state.checkpoints),
            size=self._state.size,
            session_id=self._state.session_id,
            stats=copy.copy(self._state.stats),
        )

    def get_session_id(self) -> str:
        """Get current session ID."""
        return self._state.session_id

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _evict_stale(self) -> None:
        """Evict checkpoints older than max_age_seconds."""
        now = time.time()
        max_age = self._config.max_age_seconds
        self._state.checkpoints = [
            cp
            for cp in self._state.checkpoints
            if now - _parse_timestamp(cp.timestamp) <= max_age
        ]

    def _recompute_stats(self) -> None:
        """Recompute window-level stats from current checkpoints."""
        cps = self._state.checkpoints
        self._state.size = len(cps)
        self._state.stats.clear_count = 0
        self._state.stats.review_count = 0
        self._state.stats.violation_count = 0

        total_ms = 0.0
        for cp in cps:
            if cp.verdict == "clear":
                self._state.stats.clear_count += 1
            elif cp.verdict == "review_needed":
                self._state.stats.review_count += 1
            elif cp.verdict == "boundary_violation":
                self._state.stats.violation_count += 1
            total_ms += cp.analysis_metadata.analysis_duration_ms

        self._state.stats.avg_analysis_ms = (
            total_ms / len(cps) if len(cps) > 0 else 0
        )


def _parse_timestamp(timestamp: str) -> float:
    """Parse an ISO 8601 timestamp to a Unix timestamp (seconds)."""
    # Handle both 'Z' suffix and '+00:00' offset formats
    ts = timestamp.replace("Z", "+00:00")
    return datetime.fromisoformat(ts).timestamp()
